### PR TITLE
[LM-9681] Added Topic Feed

### PR DESCRIPTION
**Pull request by ishaan1607**
## Description

**Features**
- Added Topic Feed

**Bugs**

**What kind of change does this PR introduce?** (check at least one)

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Hotfix (fix for breaking changes)
- [x] New Feature (non-breaking change which adds functionality)
- [ ] Other, please describe:

### Author Checklist

- [x] Assignee added
- [x] Reviewer added
- [x] Description added
- [ ] Unit test cases added

### Reviewer Checklist

- [x] PR title starts with Jira ticket number prefix
- [ ] If hotfix, Add Hotfix as prefix in PR title
- [x] Label added
- [x] Commit messages follows standard format (added, changed, removed, code commented and chore)
- [x] Description about the Feature/ Bugs added

### CODE REVIEW

File: feedsx/src/main/java/com/likeminds/feedsx/post/edit/view/LMFeedEditPostFragment.kt [hunk 1]
```diff
@@ -207,18 +223,76 @@ class EditPostFragment :
...
-                    ogTags = ogTags
+                    ogTags = ogTags,
+                    selectedTopics = topics.toList()
+                )
+            } else {
+                //show dialog for disabled topics
+                showDisabledTopicsAlert(disabledTopics.values.toList())
+            }
+        } else {
+            //call api as all topics are enabled
```
Comment (sidcr7-likeminds) [inline feedsx/src/main/java/com/likeminds/feedsx/post/edit/view/LMFeedEditPostFragment.kt:252]: edit comment - call api as no topics are selected

File: feedsx/src/main/java/com/likeminds/feedsx/post/edit/view/LMFeedEditPostFragment.kt [hunk 2]
```diff
@@ -207,18 +223,76 @@ class EditPostFragment :
...
-        binding.apply {
-            tvSave.setOnClickListener {
-                val text = etPostContent.text
-                val updatedText = memberTagging.replaceSelectedMembers(text).trim()
+        binding.tvSave.setOnClickListener {
+            savePost()
+        }
+    }
+
+    private fun savePost() {
```
Comment (sidcr7-likeminds) [inline feedsx/src/main/java/com/likeminds/feedsx/post/edit/view/LMFeedEditPostFragment.kt:231]: add comment

File: feedsx/src/main/java/com/likeminds/feedsx/post/edit/view/LMFeedEditPostFragment.kt [hunk 3]
```diff
@@ -207,18 +223,76 @@ class EditPostFragment :
...
+                editPostExtras.postId,
+                updatedText,
+                attachments = fileAttachments,
+                ogTags = ogTags,
+                selectedTopics = topics.toList()
+            )
+        }
+    }
+
+    private fun showDisabledTopicsAlert(disabledTopics: List<LMFeedTopicViewData>) {
```
Comment (sidcr7-likeminds) [inline feedsx/src/main/java/com/likeminds/feedsx/post/edit/view/LMFeedEditPostFragment.kt:264]: add comment

File: feedsx/src/main/java/com/likeminds/feedsx/post/edit/viewmodel/LMFeedEditPostViewModel.kt [hunk 4]
```diff
@@ -3,7 +3,10 @@ package com.likeminds.feedsx.post.edit.viewmodel
 import androidx.lifecycle.ViewModel
 import androidx.lifecycle.viewModelScope
 import com.likeminds.feedsx.LMFeedAnalytics
-import com.likeminds.feedsx.posttypes.model.*
+import com.likeminds.feedsx.posttypes.model.AttachmentViewData
+import com.likeminds.feedsx.posttypes.model.LinkOGTagsViewData
+import com.likeminds.feedsx.posttypes.model.PostViewData
```
Comment (sidcr7-likeminds) [inline feedsx/src/main/java/com/likeminds/feedsx/post/edit/viewmodel/LMFeedEditPostViewModel.kt:8]: minimise import statements

File: feedsx/src/main/java/com/likeminds/feedsx/post/edit/viewmodel/LMFeedHelperViewModel.kt [hunk 5]
```diff
@@ -12,12 +15,16 @@ import com.likeminds.feedsx.utils.membertagging.model.UserTagViewData
 import com.likeminds.feedsx.utils.membertagging.util.MemberTaggingUtil
 import com.likeminds.likemindsfeed.LMFeedClient
 import com.likeminds.likemindsfeed.LMResponse
-import com.likeminds.likemindsfeed.helper.model.*
+import com.likeminds.likemindsfeed.helper.model.DecodeUrlRequest
+import com.likeminds.likemindsfeed.helper.model.DecodeUrlResponse
+import com.likeminds.likemindsfeed.helper.model.GetTaggingListRequest
+import com.likeminds.likemindsfeed.helper.model.GetTaggingListResponse
+import com.likeminds.likemindsfeed.topic.model.GetTopicRequest
```
Comment (sidcr7-likeminds) [inline feedsx/src/main/java/com/likeminds/feedsx/post/edit/viewmodel/LMFeedHelperViewModel.kt:22]: minimise import statements

File: feedsx/src/main/java/com/likeminds/feedsx/post/edit/viewmodel/LMFeedHelperViewModel.kt [hunk 6]
```diff
@@ -119,6 +130,35 @@ class HelperViewModel @Inject constructor(
         }
     }
 
+    fun getAllTopics(showEnabledTopicsOnly: Boolean) {
```
Comment (sidcr7-likeminds) [inline feedsx/src/main/java/com/likeminds/feedsx/post/edit/viewmodel/LMFeedHelperViewModel.kt:133]: add comments

File: feedsx/src/main/java/com/likeminds/feedsx/topic/view/LMFeedTopicSelectionFragment.kt [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+            //send result
+            val resultIntent = Intent().apply {
+                putExtra(TOPIC_SELECTION_RESULT_EXTRAS, resultExtras)
+            }
+            requireActivity().setResult(Activity.RESULT_OK, resultIntent)
+            requireActivity().finish()
+        }
+    }
+
+    private fun checkForSelectedTopics() {
```
Comment (sidcr7-likeminds) [inline feedsx/src/main/java/com/likeminds/feedsx/topic/view/LMFeedTopicSelectionFragment.kt:226]: add comments

File: feedsx/src/main/java/com/likeminds/feedsx/topic/viewmodel/LMFeedTopicSelectionViewModel.kt [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+    }
+
+    fun setPreviousSelectedTopics(selectedTopics: List<LMFeedTopicViewData>?) {
+        if (selectedTopics.isNullOrEmpty()) return
+        selectedTopics.forEach { topic ->
+            this.selectedTopics[topic.id] = topic
+        }
+    }
+
+    fun setPreviousDisabledTopics(disabledTopics: List<LMFeedTopicViewData>?) {
```
Comment (sidcr7-likeminds) [inline feedsx/src/main/java/com/likeminds/feedsx/topic/viewmodel/LMFeedTopicSelectionViewModel.kt:47]: add comments

File: feedsx/src/main/java/com/likeminds/feedsx/topic/viewmodel/LMFeedTopicSelectionViewModel.kt [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+        if (disabledTopics.isNullOrEmpty()) return
+        val updatedDisabledTopics = disabledTopics.map {
+            it.toBuilder().isSelected(true).build()
+        }
+        this.disabledTopics.addAll(updatedDisabledTopics)
+    }
+
+    fun getTopics(
```
Comment (sidcr7-likeminds) [inline feedsx/src/main/java/com/likeminds/feedsx/topic/viewmodel/LMFeedTopicSelectionViewModel.kt:55]: add comments

File: feedsx/src/main/java/com/likeminds/feedsx/topic/viewmodel/LMFeedTopicSelectionViewModel.kt [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+                .pageSize(PAGE_SIZE)
+
+            if (!searchString.isNullOrEmpty()) {
+                requestBuilder.search(searchString)
+                    .searchType(SEARCH_TYPE)
+            }
+
+            if (showEnabledTopicOnly) {
+                requestBuilder.isEnabled(true)
+            }
```
Comment (sidcr7-likeminds) [inline feedsx/src/main/java/com/likeminds/feedsx/topic/viewmodel/LMFeedTopicSelectionViewModel.kt:74]: is this check required? why not send `showEnabledTopicOnly` as a variable

**Review by sidcr7-likeminds — CHANGES_REQUESTED** [review]

File: feedsx/src/main/java/com/likeminds/feedsx/feed/viewmodel/FeedViewModel.kt [hunk 11]
```diff
@@ -207,6 +207,35 @@ class FeedViewModel @Inject constructor(
         }
     }
 
+    fun getAllTopics() {
```
[the remark this answers is not in the document]
Comment (ishaan1607) [reply feedsx/src/main/java/com/likeminds/feedsx/feed/viewmodel/FeedViewModel.kt]: this function is moved to `HelperViewModel`

File: feedsx/src/main/java/com/likeminds/feedsx/post/create/viewmodel/CreatePostViewModel.kt [hunk 12]
```diff
@@ -80,23 +100,40 @@ class CreatePostViewModel @Inject constructor(
...
                 if (ogTags != null) {
                     // if the post has ogTags
                     requestBuilder.attachments(ViewDataConverter.convertAttachments(ogTags))
                 }
+
+                if (!topicIds.isNullOrEmpty()) {
+                    //if user has selected any topics
+                    requestBuilder.topicIds(topicIds)
+                }
+
```
[the remark this answers is not in the document]
Comment (ishaan1607) [reply feedsx/src/main/java/com/likeminds/feedsx/post/create/viewmodel/CreatePostViewModel.kt:108]: ignored

File: feedsx/src/main/java/com/likeminds/feedsx/topic/viewmodel/LMFeedTopicSelectionViewModel.kt (hunk 10 above)
Comment (ishaan1607) [reply feedsx/src/main/java/com/likeminds/feedsx/topic/viewmodel/LMFeedTopicSelectionViewModel.kt:74]: Yes, it is required

**Review by sidcr7-likeminds — APPROVED** [review]

**Pull request merged**